File: packages/nexus-coordinator/src/nexus_coordinator/canary_registry.py
# SPDX-License-Identifier: AGPL-3.0-or-later
"""Sprint 20 Phase E.3 — federated warrant canary registry.

Aggregates every signed warrant canary the coordinator has
observed (typically from peer-side gossip subscribe + local CLI
publish, but the source is opaque to the registry) and exposes
freshness state per maintainer pubkey.

## Shape

The registry is keyed by maintainer **pubkey** (lowercase hex of
the 32-byte Ed25519 verifying key embedded in
:class:`CanarySigned`) and stores the most recent canary date /
headline / next-update / signature seen for that pubkey, plus the
most recent duress ack date observed (Phase E.4 channel — same
maintainer, separate cadence).

## Wire-format invariants honoured

- The registry never re-signs anything — it just records what it
  observed. Signature verification is the responsibility of the
  caller (`verify_canary` / `verify_duress_ack` Rust path,
  surfaced over HTTP by the daemon side later).
- The pubkey strings are stored verbatim from the signed payload,
  64 lowercase hex chars. The registry does no normalisation.
- The wire format embedded in the persisted JSON is the same
  ``{ signed: CanarySigned, signature_hex }`` shape the daemon
  emits — verifiers can re-verify offline by re-deriving the
  canonical bytes (cf. ``DOMAIN_WARRANT_CANARY_V1``).

## Threat model coverage

Phase E.3 is the **observability** layer. It does NOT make any
trust decision on its own — a stale registry is just data, the
operator interprets it. By exposing the freshness through
``GET /api/canary/network-health``, the shell can render a fleet
view ("3 maintainers all fresh / 1 maintainer 38 days stale, last
ack 4 days ago") that gives a human a single pane to spot a
coercion pattern earlier than re-deriving it from raw gossip.

## Persistence

The registry persists to a single JSON file at
:func:`canary_registry_path` on every observation. The format is
forward-compatible-by-construction (new fields are added at the
top level with sensible defaults, never inside the per-pubkey
objects which mirror the wire format byte-for-byte).
"""

from __future__ import annotations

import json
from collections.abc import Iterable
from datetime import date, datetime, timezone
from pathlib import Path
from typing import Any

import structlog
from pydantic import BaseModel, ConfigDict, Field

_log = structlog.get_logger(__name__)

# Mirror of the Rust :rust:`CANARY_VALIDITY_DAYS` constant.
# Keep these two in sync — the freshness alarm thresholds derive
# from the canary's own validity window.
CANARY_VALIDITY_DAYS: int = 45

# Freshness alarm thresholds. A canary is "fresh" if its date is
# within the validity window, "warn" if it has overshot the window
# but stays within the soft tail, and "stale" past the soft tail.
# These match the Sprint 18 Phase E2 design : 30-day cadence,
# 15-day grace, 30-day soft tail before fully alarming.
WARN_THRESHOLD_DAYS: int = 30  # cadence ; canary is expected every 30 days
ALARM_THRESHOLD_DAYS: int = CANARY_VALIDITY_DAYS  # 45 ; past this = dead-man-switch fired

# Duress acks have a much tighter cadence (daily-or-better).
# Past 2 days without an ack is the operator's first signal.
DURESS_ACK_WARN_DAYS: int = 2
DURESS_ACK_ALARM_DAYS: int = 7


class CanaryObservation(BaseModel):
    """The signed body + signature the registry persists per pubkey.

    Mirrors the Rust :rust:`Canary` wire shape exactly. Fields are
    typed permissively (``str`` for the date / pubkey / sig hex)
    because the Python side does not re-validate the cryptography
    — it stores what it observed and re-emits it verbatim.
    """

    model_config = ConfigDict(extra="forbid")

    version: int = Field(..., description="CANARY_VERSION constant ; pre-launch always 1.")
    date: str = Field(..., description="UTC date YYYY-MM-DD of the canary.")
    headline: str
    next_update: str
    pubkey_hex: str = Field(..., min_length=64, max_length=64)
    signature_hex: str = Field(..., min_length=128, max_length=128)


class DuressAckObservation(BaseModel):
    """The signed body + signature of an observed duress ack.

    Mirrors the Rust :rust:`DuressAck` wire shape.
    """

    model_config = ConfigDict(extra="forbid")

    version: int = Field(..., description="DURESS_ACK_VERSION constant ; pre-launch always 1.")
    date: str = Field(..., description="UTC date YYYY-MM-DD of the ack.")
    message: str
    pubkey_hex: str = Field(..., min_length=64, max_length=64)
    signature_hex: str = Field(..., min_length=128, max_length=128)


class CanaryFreshness(BaseModel):
    """Freshness diagnostic for a single maintainer pubkey.

    The registry computes this on demand from the stored
    observation timestamps, not from the gossip event timestamps,
    so a verifier seeing the API response can reason directly about
    "the maintainer last claimed liberty on this date".
    """

    model_config = ConfigDict(extra="forbid")

    pubkey_hex: str
    canary_date: str | None = Field(default=None, description="Most recent canary date observed.")
    canary_age_days: int | None = Field(default=None, ge=0)
    canary_status: str = Field(
        default="missing",
        description="One of fresh / warn / stale / missing.",
    )
    duress_ack_date: str | None = Field(default=None, description="Most recent duress ack observed.")
    duress_ack_age_days: int | None = Field(default=None, ge=0)
    duress_ack_status: str = Field(
        default="missing",
        description="One of fresh / warn / stale / missing.",
    )


class NetworkHealth(BaseModel):
    """Top-level shape returned by ``GET /api/canary/network-health``.

    Designed for direct consumption by the React shell — the
    counts in ``summary`` give a one-glance fleet picture, the
    per-maintainer ``maintainers`` array drives a detail view.
    """

    model_config = ConfigDict(extra="forbid")

    summary: dict[str, int]
    maintainers: list[CanaryFreshness]
    observed_at: str = Field(..., description="UTC RFC 3339 timestamp of this snapshot.")


def _today_utc() -> date:
    return datetime.now(timezone.utc).date()


def _classify_canary_age(age_days: int) -> str:
    if age_days < WARN_THRESHOLD_DAYS:
        return "fresh"
    if age_days < ALARM_THRESHOLD_DAYS:
        return "warn"
    return "stale"


def _classify_duress_age(age_days: int) -> str:
    if age_days < DURESS_ACK_WARN_DAYS:
        return "fresh"
    if age_days < DURESS_ACK_ALARM_DAYS:
        return "warn"
    return "stale"


class CanaryRegistry:
    """In-memory + on-disk aggregator of observed warrant canaries.

    Thread safety : single-threaded by construction. The
    coordinator process is asyncio-driven and the registry is
    only mutated from FastAPI request handlers and the
    coordinator's own boot path, never from a thread pool.
    """

    def __init__(self, persist_path: Path) -> None:
        self._persist_path = persist_path
        self._canaries: dict[str, CanaryObservation] = {}
        self._duress_acks: dict[str, DuressAckObservation] = {}
        self._load_if_exists()

    # ------------------------------------------------------------------
    # Persistence
    # ------------------------------------------------------------------

    def _load_if_exists(self) -> None:
        if not self._persist_path.exists():
            return
        try:
            raw = json.loads(self._persist_path.read_text(encoding="utf-8"))
        except (OSError, json.JSONDecodeError) as exc:
            _log.warning(
                "canary_registry.load_failed",
                path=str(self._persist_path),
                error=str(exc),
            )
            return
        for item in raw.get("canaries", []):
            try:
                obs = CanaryObservation.model_validate(item)
            except Exception as exc:  # pydantic ValidationError or any unexpected shape
                _log.warning("canary_registry.bad_canary_entry", error=str(exc))
                continue
            self._canaries[obs.pubkey_hex] = obs
        for item in raw.get("duress_acks", []):
            try:
                obs = DuressAckObservation.model_validate(item)
            except Exception as exc:
                _log.warning("canary_registry.bad_duress_ack_entry", error=str(exc))
                continue
            self._duress_acks[obs.pubkey_hex] = obs

    def _persist(self) -> None:
        self._persist_path.parent.mkdir(parents=True, exist_ok=True)
        payload = {
            "canaries": [obs.model_dump() for obs in self._canaries.values()],
            "duress_acks": [obs.model_dump() for obs in self._duress_acks.values()],
        }
        # Atomic write : write to a temp file then rename so a
        # crashed coordinator never leaves a half-written JSON
        # blob behind.
        tmp = self._persist_path.with_suffix(self._persist_path.suffix + ".tmp")
        tmp.write_text(json.dumps(payload, indent=2, sort_keys=True), encoding="utf-8")
        tmp.replace(self._persist_path)

    # ------------------------------------------------------------------
    # Observation
    # ------------------------------------------------------------------

    def observe_canary(self, observation: CanaryObservation) -> None:
        """Record a freshly observed warrant canary.

        Only the most recent canary per pubkey is kept (by ``date``
        string lexicographic order, which is a valid total order
        for ISO-8601 ``YYYY-MM-DD``). An older observation for the
        same pubkey is silently ignored.
        """
        prev = self._canaries.get(observation.pubkey_hex)
        if prev is not None and prev.date >= observation.date:
            return
        self._canaries[observation.pubkey_hex] = observation
        self._persist()

    def observe_duress_ack(self, observation: DuressAckObservation) -> None:
        """Record a freshly observed duress ack."""
        prev = self._duress_acks.get(observation.pubkey_hex)
        if prev is not None and prev.date >= observation.date:
            return
        self._duress_acks[observation.pubkey_hex] = observation
        self._persist()

    # ------------------------------------------------------------------
    # Inspection
    # ------------------------------------------------------------------

    def known_pubkeys(self) -> Iterable[str]:
        """Pubkeys with at least one observation (canary or ack)."""
        return set(self._canaries) | set(self._duress_acks)

    def freshness(self, pubkey_hex: str, today: date | None = None) -> CanaryFreshness:
        """Compute the freshness diagnostic for a single pubkey."""
        today = today or _today_utc()
        result = CanaryFreshness(pubkey_hex=pubkey_hex)

        canary = self._canaries.get(pubkey_hex)
        if canary is not None:
            result.canary_date = canary.date
            try:
                age = (today - date.fromisoformat(canary.date)).days
            except ValueError:
                age = None
            if age is not None and age >= 0:
                result.canary_age_days = age
                result.canary_status = _classify_canary_age(age)

        ack = self._duress_acks.get(pubkey_hex)
        if ack is not None:
            result.duress_ack_date = ack.date
            try:
                age = (today - date.fromisoformat(ack.date)).days
            except ValueError:
                age = None
            if age is not None and age >= 0:
                result.duress_ack_age_days = age
                result.duress_ack_status = _classify_duress_age(age)

        return result

    def network_health(self, today: date | None = None) -> NetworkHealth:
        """Snapshot of every known maintainer's freshness."""
        today = today or _today_utc()
        per_pubkey = [self.freshness(pk, today=today) for pk in sorted(self.known_pubkeys())]

        summary: dict[str, int] = {
            "maintainers_total": len(per_pubkey),
            "canary_fresh": 0,
            "canary_warn": 0,
            "canary_stale": 0,
            "canary_missing": 0,
            "duress_ack_fresh": 0,
            "duress_ack_warn": 0,
            "duress_ack_stale": 0,
            "duress_ack_missing": 0,
        }
        for entry in per_pubkey:
            summary[f"canary_{entry.canary_status}"] += 1
            summary[f"duress_ack_{entry.duress_ack_status}"] += 1

        return NetworkHealth(
            summary=summary,
            maintainers=per_pubkey,
            observed_at=datetime.now(timezone.utc).isoformat(timespec="seconds"),
        )

    # ------------------------------------------------------------------
    # Test affordance
    # ------------------------------------------------------------------

    def _canaries_snapshot(self) -> dict[str, CanaryObservation]:
        """Read-only snapshot for tests."""
        return dict(self._canaries)

    def _duress_acks_snapshot(self) -> dict[str, DuressAckObservation]:
        """Read-only snapshot for tests."""
        return dict(self._duress_acks)


def coerce_canary_payload(payload: dict[str, Any]) -> CanaryObservation:
    """Build a :class:`CanaryObservation` from the daemon's wire payload.

    The Rust :rust:`Canary` struct serializes to a flat object
    via ``#[serde(flatten)]`` on the inner ``signed`` field, so
    the on-the-wire JSON is::

        {
            "v": 1, "date": "...", "headline": "...",
            "next_update": "...", "pubkey_hex": "...",
            "signature_hex": "..."
        }

    This helper handles the ``v`` -> ``version`` rename and tolerates
    payloads that already use the explicit ``version`` key (Python
    test fixtures, manual ``POST`` bodies).
    """
    body = dict(payload)
    if "v" in body and "version" not in body:
        body["version"] = body.pop("v")
    return CanaryObservation.model_validate(body)


def coerce_duress_ack_payload(payload: dict[str, Any]) -> DuressAckObservation:
    """Same as :func:`coerce_canary_payload` for duress acks."""
    body = dict(payload)
    if "v" in body and "version" not in body:
        body["version"] = body.pop("v")
    return DuressAckObservation.model_validate(body)

File: crates/nexus-coordinator-rs/src/canary_registry.rs
// SPDX-License-Identifier: AGPL-3.0-or-later
//! Warrant canary registry — aggregates observed canary signings and
//! duress acks, computes freshness per maintainer pubkey.
//!
//! Port of `packages/nexus-coordinator/src/nexus_coordinator/canary_registry.py`
//! (366 LOC Python → Rust, Sprint 39 Phase B).

use std::collections::HashMap;
use std::io;
use std::path::PathBuf;

use serde::{Deserialize, Serialize};
use time::{Date, OffsetDateTime};

const WARN_THRESHOLD_DAYS: i64 = 30;
const ALARM_THRESHOLD_DAYS: i64 = 45;
const DURESS_ACK_WARN_DAYS: i64 = 2;
const DURESS_ACK_ALARM_DAYS: i64 = 7;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CanaryObservation {
    pub version: u32,
    pub date: String,
    pub headline: String,
    pub next_update: String,
    pub pubkey_hex: String,
    pub signature_hex: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct DuressAckObservation {
    pub version: u32,
    pub date: String,
    pub message: String,
    pub pubkey_hex: String,
    pub signature_hex: String,
}

#[derive(Debug, Clone, Default, Serialize)]
pub struct CanaryFreshness {
    pub pubkey_hex: String,
    pub canary_date: Option<String>,
    pub canary_age_days: Option<i64>,
    pub canary_status: String,
    pub duress_ack_date: Option<String>,
    pub duress_ack_age_days: Option<i64>,
    pub duress_ack_status: String,
}

#[derive(Debug, Clone, Serialize)]
pub struct NetworkHealth {
    pub summary: HashMap<String, i64>,
    pub maintainers: Vec<CanaryFreshness>,
    pub observed_at: String,
}

fn classify_canary_age(days: i64) -> &'static str {
    if days < WARN_THRESHOLD_DAYS {
        "fresh"
    } else if days < ALARM_THRESHOLD_DAYS {
        "warn"
    } else {
        "stale"
    }
}

fn classify_duress_age(days: i64) -> &'static str {
    if days < DURESS_ACK_WARN_DAYS {
        "fresh"
    } else if days < DURESS_ACK_ALARM_DAYS {
        "warn"
    } else {
        "stale"
    }
}

fn today_utc() -> Date {
    OffsetDateTime::now_utc().date()
}

fn parse_date(s: &str) -> Option<Date> {
    let format = time::macros::format_description!("[year]-[month]-[day]");
    Date::parse(s, &format).ok()
}

#[derive(Debug, Default, Serialize, Deserialize)]
struct PersistedRegistry {
    canaries: Vec<CanaryObservation>,
    duress_acks: Vec<DuressAckObservation>,
}

#[derive(Debug)]
pub struct CanaryRegistry {
    canaries: HashMap<String, CanaryObservation>,
    duress_acks: HashMap<String, DuressAckObservation>,
    persist_path: PathBuf,
}

impl CanaryRegistry {
    pub fn new(persist_path: PathBuf) -> Self {
        let mut reg = Self {
            canaries: HashMap::new(),
            duress_acks: HashMap::new(),
            persist_path,
        };
        reg.load_if_exists();
        reg
    }

    fn load_if_exists(&mut self) {
        if !self.persist_path.exists() {
            return;
        }
        let text = match std::fs::read_to_string(&self.persist_path) {
            Ok(t) => t,
            Err(e) => {
                tracing::warn!(path = %self.persist_path.display(), error = %e, "canary_registry.load_failed");
                return;
            }
        };
        let raw: PersistedRegistry = match serde_json::from_str(&text) {
            Ok(r) => r,
            Err(e) => {
                tracing::warn!(path = %self.persist_path.display(), error = %e, "canary_registry.parse_failed");
                return;
            }
        };
        for obs in raw.canaries {
            self.canaries.insert(obs.pubkey_hex.clone(), obs);
        }
        for obs in raw.duress_acks {
            self.duress_acks.insert(obs.pubkey_hex.clone(), obs);
        }
    }

    pub fn persist(&self) -> io::Result<()> {
        if let Some(parent) = self.persist_path.parent() {
            std::fs::create_dir_all(parent)?;
        }
        let payload = PersistedRegistry {
            canaries: self.canaries.values().cloned().collect(),
            duress_acks: self.duress_acks.values().cloned().collect(),
        };
        let json = serde_json::to_string_pretty(&payload).map_err(io::Error::other)?;
        let tmp = self.persist_path.with_extension("json.tmp");
        std::fs::write(&tmp, &json)?;
        std::fs::rename(&tmp, &self.persist_path)?;
        Ok(())
    }

    pub fn observe_canary(&mut self, obs: CanaryObservation) {
        if let Some(prev) = self.canaries.get(&obs.pubkey_hex) {
            if prev.date >= obs.date {
                return;
            }
        }
        self.canaries.insert(obs.pubkey_hex.clone(), obs);
        let _ = self.persist();
    }

    pub fn observe_duress_ack(&mut self, obs: DuressAckObservation) {
        if let Some(prev) = self.duress_acks.get(&obs.pubkey_hex) {
            if prev.date >= obs.date {
                return;
            }
        }
        self.duress_acks.insert(obs.pubkey_hex.clone(), obs);
        let _ = self.persist();
    }

    pub fn known_pubkeys(&self) -> Vec<String> {
        let mut keys: std::collections::HashSet<&str> =
            self.canaries.keys().map(|k| k.as_str()).collect();
        for k in self.duress_acks.keys() {
            keys.insert(k.as_str());
        }
        let mut sorted: Vec<String> = keys.into_iter().map(|s| s.to_string()).collect();
        sorted.sort();
        sorted
    }

    pub fn freshness(&self, pubkey_hex: &str) -> CanaryFreshness {
        self.freshness_at(pubkey_hex, today_utc())
    }

    pub fn freshness_at(&self, pubkey_hex: &str, today: Date) -> CanaryFreshness {
        let mut result = CanaryFreshness {
            pubkey_hex: pubkey_hex.to_string(),
            canary_status: "missing".into(),
            duress_ack_status: "missing".into(),
            ..Default::default()
        };

        if let Some(canary) = self.canaries.get(pubkey_hex) {
            result.canary_date = Some(canary.date.clone());
            if let Some(d) = parse_date(&canary.date) {
                let age = (today - d).whole_days();
                if age >= 0 {
                    result.canary_age_days = Some(age);
                    result.canary_status = classify_canary_age(age).into();
                }
            }
        }

        if let Some(ack) = self.duress_acks.get(pubkey_hex) {
            result.duress_ack_date = Some(ack.date.clone());
            if let Some(d) = parse_date(&ack.date) {
                let age = (today - d).whole_days();
                if age >= 0 {
                    result.duress_ack_age_days = Some(age);
                    result.duress_ack_status = classify_duress_age(age).into();
                }
            }
        }

        result
    }

    pub fn network_health(&self) -> NetworkHealth {
        self.network_health_at(today_utc())
    }

    pub fn network_health_at(&self, today: Date) -> NetworkHealth {
        let pubkeys = self.known_pubkeys();
        let maintainers: Vec<CanaryFreshness> = pubkeys
            .iter()
            .map(|pk| self.freshness_at(pk, today))
            .collect();

        let mut summary = HashMap::new();
        summary.insert("maintainers_total".into(), maintainers.len() as i64);
        for status in ["fresh", "warn", "stale", "missing"] {
            summary.insert(format!("canary_{status}"), 0);
            summary.insert(format!("duress_ack_{status}"), 0);
        }
        for entry in &maintainers {
            *summary
                .entry(format!("canary_{}", entry.canary_status))
                .or_insert(0) += 1;
            *summary
                .entry(format!("duress_ack_{}", entry.duress_ack_status))
                .or_insert(0) += 1;
        }

        NetworkHealth {
            summary,
            maintainers,
            observed_at: OffsetDateTime::now_utc()
                .format(&time::macros::format_description!(
                    "[year]-[month]-[day]T[hour]:[minute]:[second]Z"
                ))
                .unwrap_or_default(),
        }
    }
}

pub fn coerce_canary_payload(payload: &serde_json::Value) -> Result<CanaryObservation, String> {
    let mut obj = match payload.as_object() {
        Some(o) => o.clone(),
        None => return Err("expected JSON object".into()),
    };
    if obj.contains_key("v") && !obj.contains_key("version") {
        if let Some(v) = obj.remove("v") {
            obj.insert("version".into(), v);
        }
    }
    serde_json::from_value(serde_json::Value::Object(obj)).map_err(|e| e.to_string())
}

pub fn coerce_duress_ack_payload(
    payload: &serde_json::Value,
) -> Result<DuressAckObservation, String> {
    let mut obj = match payload.as_object() {
        Some(o) => o.clone(),
        None => return Err("expected JSON object".into()),
    };
    if obj.contains_key("v") && !obj.contains_key("version") {
        if let Some(v) = obj.remove("v") {
            obj.insert("version".into(), v);
        }
    }
    serde_json::from_value(serde_json::Value::Object(obj)).map_err(|e| e.to_string())
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::TempDir;

    fn make_canary(pubkey: &str, date: &str) -> CanaryObservation {
        CanaryObservation {
            version: 1,
            date: date.into(),
            headline: "all clear".into(),
            next_update: "2026-06-01".into(),
            pubkey_hex: pubkey.into(),
            signature_hex: "a".repeat(128),
        }
    }

    fn make_ack(pubkey: &str, date: &str) -> DuressAckObservation {
        DuressAckObservation {
            version: 1,
            date: date.into(),
            message: "ack".into(),
            pubkey_hex: pubkey.into(),
            signature_hex: "b".repeat(128),
        }
    }

    #[test]
    fn observe_canary_keeps_latest() {
        let dir = TempDir::new().unwrap();
        let path = dir.path().join("canary.json");
        let mut reg = CanaryRegistry::new(path);
        let pk = "a".repeat(64);
        reg.observe_canary(make_canary(&pk, "2026-04-01"));
        reg.observe_canary(make_canary(&pk, "2026-04-15"));
        reg.observe_canary(make_canary(&pk, "2026-04-10"));
        assert_eq!(reg.canaries[&pk].date, "2026-04-15");
    }

    #[test]
    fn observe_duress_ack_keeps_latest() {
        let dir = TempDir::new().unwrap();
        let path = dir.path().join("canary.json");
        let mut reg = CanaryRegistry::new(path);
        let pk = "b".repeat(64);
        reg.observe_duress_ack(make_ack(&pk, "2026-04-01"));
        reg.observe_duress_ack(make_ack(&pk, "2026-04-20"));
        assert_eq!(reg.duress_acks[&pk].date, "2026-04-20");
    }

    #[test]
    fn freshness_fresh() {
        let dir = TempDir::new().unwrap();
        let mut reg = CanaryRegistry::new(dir.path().join("c.json"));
        let pk = "c".repeat(64);
        reg.observe_canary(make_canary(&pk, "2026-04-25"));
        let today = Date::from_calendar_date(2026, time::Month::April, 29).unwrap();
        let f = reg.freshness_at(&pk, today);
        assert_eq!(f.canary_status, "fresh");
        assert_eq!(f.canary_age_days, Some(4));
    }

    #[test]
    fn freshness_stale() {
        let dir = TempDir::new().unwrap();
        let mut reg = CanaryRegistry::new(dir.path().join("c.json"));
        let pk = "d".repeat(64);
        reg.observe_canary(make_canary(&pk, "2026-03-01"));
        let today = Date::from_calendar_date(2026, time::Month::April, 29).unwrap();
        let f = reg.freshness_at(&pk, today);
        assert_eq!(f.canary_status, "stale");
        assert_eq!(f.canary_age_days, Some(59));
    }

    #[test]
    fn freshness_unknown_key() {
        let dir = TempDir::new().unwrap();
        let reg = CanaryRegistry::new(dir.path().join("c.json"));
        let f = reg.freshness("unknown");
        assert_eq!(f.canary_status, "missing");
        assert_eq!(f.duress_ack_status, "missing");
    }

    #[test]
    fn network_health_mixed() {
        let dir = TempDir::new().unwrap();
        let mut reg = CanaryRegistry::new(dir.path().join("c.json"));
        let pk_fresh = "e".repeat(64);
        let pk_stale = "f".repeat(64);
        let pk_ack_only = "0".repeat(64);
        reg.observe_canary(make_canary(&pk_fresh, "2026-04-25"));
        reg.observe_canary(make_canary(&pk_stale, "2026-03-01"));
        reg.observe_duress_ack(make_ack(&pk_ack_only, "2026-04-28"));
        let today = Date::from_calendar_date(2026, time::Month::April, 29).unwrap();
        let h = reg.network_health_at(today);
        assert_eq!(h.summary["maintainers_total"], 3);
        assert_eq!(h.summary["canary_fresh"], 1);
        assert_eq!(h.summary["canary_stale"], 1);
        assert_eq!(h.summary["canary_missing"], 1);
        assert_eq!(h.summary["duress_ack_fresh"], 1);
        assert_eq!(h.summary["duress_ack_missing"], 2);
    }

    #[test]
    fn persist_and_reload() {
        let dir = TempDir::new().unwrap();
        let path = dir.path().join("canary.json");
        let pk = "1".repeat(64);
        {
            let mut reg = CanaryRegistry::new(path.clone());
            reg.observe_canary(make_canary(&pk, "2026-04-20"));
            reg.observe_duress_ack(make_ack(&pk, "2026-04-21"));
        }
        let reg2 = CanaryRegistry::new(path);
        assert_eq!(reg2.canaries[&pk].date, "2026-04-20");
        assert_eq!(reg2.duress_acks[&pk].date, "2026-04-21");
    }

    #[test]
    fn coerce_canary_payload_v_rename() {
        let json = serde_json::json!({
            "v": 1,
            "date": "2026-04-20",
            "headline": "ok",
            "next_update": "2026-05-20",
            "pubkey_hex": "a".repeat(64),
            "signature_hex": "b".repeat(128),
        });
        let obs = coerce_canary_payload(&json).unwrap();
        assert_eq!(obs.version, 1);
        assert_eq!(obs.date, "2026-04-20");
    }

    #[test]
    fn coerce_canary_payload_missing_field() {
        let json = serde_json::json!({"v": 1, "date": "2026-04-20"});
        assert!(coerce_canary_payload(&json).is_err());
    }
}
